lib/spack/spack/solver/input_analysis.py

```diff
@@ -0,0 +1,524 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Classes to analyze the input of a solve, and provide information to set up the ASP problem"""
+import collections
+from typing import Dict, List, NamedTuple, Set, Tuple, Union
+
+import archspec.cpu
+
+from llnl.util import lang, tty
+
+import spack.binary_distribution
+import spack.config
+import spack.deptypes as dt
+import spack.platforms
+import spack.repo
+import spack.spec
+import spack.store
+from spack.error import SpackError
+
+RUNTIME_TAG = "runtime"
+
+
+class PossibleGraph(NamedTuple):
+    real_pkgs: Set[str]
+    virtuals: Set[str]
+    edges: Dict[str, Set[str]]
+
+
+class PossibleDependencyGraph:
```

Comment:
What I meant was to have a base class with an implementation for traversal of the graph, and either subclass it or do a visitor pattern for static analysis where you don't visit certain edges. Is that not feasible? Now `possible_dependencies` is duplicated again, and calling `create_graph_analyzer` from `graph.py` and `cmd/dependencies.py` is incorrect since they should be static? Or is that intentional?

Reply:
> Now possible_dependencies is coupled to "static analysis", and calling create_graph_analyzer from graph.py and cmd/dependencies.py is incorrect since they should be config independent? Or is that intentional?

It's intentional, since it's a quick way to show what is considered during the solve, given the current configuration. For instance, with:
```yaml
packages:
  mpi:
    require: mpich
  java:
    require: openjdk
  zlib-api:
    require: zlib-ng+compat
```
you should get something like:
```
$ spack -c concretizer:static_analysis:false dependencies -t hdf5 | wc -l
563
$ spack -c concretizer:static_analysis:true dependencies -t hdf5 | wc -l
513

```

Reply:
> I think your interface is an interface for analysis, not for traversal.

Correct